File: client-android/lib/src/main/kotlin/stasis/client_android/lib/model/core/Manifest.kt
package stasis.client_android.lib.model.core

import com.squareup.moshi.JsonClass
import java.time.Instant

@JsonClass(generateAdapter = true)
data class Manifest(
    val crate: CrateId,
    val size: Long,
    val copies: Int,
    val origin: NodeId,
    val source: NodeId,
    val destinations: List<NodeId>,
    val created: Instant
) {
    companion object {
        operator fun invoke(
            crate: CrateId,
            size: Long,
            copies: Int,
            origin: NodeId,
            source: NodeId
        ): Manifest = Manifest(
            crate = crate,
            size = size,
            copies = copies,
            origin = origin,
            source = source,
            destinations = emptyList(),
            created = Instant.now()
        )
    }
}

File: client-ios/lib/Sources/StasisClientLib/Model/Core/Manifest.swift
import Foundation

public struct Manifest: Sendable, Equatable, Hashable, Codable {
    public let crate: CrateId
    public let size: Int64
    public let copies: Int
    public let origin: NodeId
    public let source: NodeId
    public let destinations: [NodeId]
    public let created: Date

    public init(
        crate: CrateId,
        size: Int64,
        copies: Int,
        origin: NodeId,
        source: NodeId,
        destinations: [NodeId],
        created: Date
    ) {
        self.crate = crate
        self.size = size
        self.copies = copies
        self.origin = origin
        self.source = source
        self.destinations = destinations
        self.created = created
    }

    public init(
        crate: CrateId,
        size: Int64,
        copies: Int,
        origin: NodeId,
        source: NodeId
    ) {
        self.init(
            crate: crate,
            size: size,
            copies: copies,
            origin: origin,
            source: source,
            destinations: [],
            created: Date()
        )
    }
}
